File: backend/src/main/java/com/reallink/pump/entities/UserAbsence.java
package com.reallink.pump.entities;

import java.time.LocalDate;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "user_absence", indexes = {
    @Index(name = "idx_user_id_absence", columnList = "user_id"),
    @Index(name = "idx_pump_master_id_absence", columnList = "pump_master_id"),
    @Index(name = "idx_absence_date", columnList = "absence_date")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class UserAbsence extends BaseEntity {

    @NotNull(message = "User is required")
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false, foreignKey = @ForeignKey(name = "fk_absence_user"))
    private User user;

    @NotNull(message = "Pump master is required")
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "pump_master_id", nullable = false, foreignKey = @ForeignKey(name = "fk_absence_pump_master"))
    private PumpInfoMaster pumpMaster;

    @NotNull(message = "Absence date is required")
    @Column(name = "absence_date", nullable = false)
    private LocalDate absenceDate;

    @Size(max = 500, message = "Reason must not exceed 500 characters")
    @Column(name = "reason", length = 500)
    private String reason;

    @Size(max = 1000, message = "Notes must not exceed 1000 characters")
    @Column(name = "notes", length = 1000)
    private String notes;

    @Column(name = "is_approved", nullable = false)
    private Boolean isApproved = false;

    @Column(name = "approved_by")
    private String approvedBy;
}

File: frontend/src/types/user-absence.ts
import { z } from "zod";

// Absence Type constant
export const AbsenceType = {
  FULL_DAY: "FULL_DAY",
  HALF_DAY: "HALF_DAY",
  OVERTIME: "OVERTIME",
} as const;

export type AbsenceType = (typeof AbsenceType)[keyof typeof AbsenceType];

// User Absence schemas
export const UserAbsenceSchema = z.object({
  id: z.string().optional(),
  userId: z.string(),
  username: z.string().optional(),
  userRole: z.string().optional(),
  pumpMasterId: z.string(),
  absenceDate: z.string(), // ISO date string
  absenceType: z
    .enum(["FULL_DAY", "HALF_DAY", "OVERTIME"])
    .default(AbsenceType.FULL_DAY),
  reason: z
    .string()
    .max(500, "Reason must not exceed 500 characters")
    .optional(),
  notes: z
    .string()
    .max(1000, "Notes must not exceed 1000 characters")
    .optional(),
  isApproved: z.boolean().default(false),
  approvedBy: z.string().optional(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
  version: z.number().optional(),
});

export const CreateUserAbsenceSchema = UserAbsenceSchema.omit({
  id: true,
  username: true,
  userRole: true,
  approvedBy: true,
  createdAt: true,
  updatedAt: true,
  version: true,
});

export const UpdateUserAbsenceSchema = z.object({
  absenceDate: z.string().optional(),
  absenceType: z.enum(["FULL_DAY", "HALF_DAY", "OVERTIME"]).optional(),
  reason: z.string().max(500).optional(),
  notes: z.string().max(1000).optional(),
  isApproved: z.boolean().optional(),
});

export type UserAbsence = z.infer<typeof UserAbsenceSchema>;
export type CreateUserAbsence = z.infer<typeof CreateUserAbsenceSchema>;
export type UpdateUserAbsence = z.infer<typeof UpdateUserAbsenceSchema>;
